### PR TITLE
Handle situation where IAM Role already exists in AWS more gracefully

### PR DESCRIPTION
**Pull request by diranged**
**What does this PR do?**

We discovered a race condition that occurs when an AWS IAM Role already exists and you create your `Iamrole` resource. The problem is that when the IAM Controller calls the [`CreateRole`](https://github.com/keikoproj/iam-manager/blob/v0.0.7/pkg/awsapi/iam.go#L66-L160) function, it tries to gracefully handle an IAM Role that already exists (see https://github.com/keikoproj/iam-manager/blob/v0.0.7/pkg/awsapi/iam.go#L103-L105) Unfortunately in that situation, it does not set the `RoleARN` or `RoleID` values (see https://github.com/keikoproj/iam-manager/blob/v0.0.7/pkg/awsapi/iam.go#L110-L113).

**What did I change?**

After digging through the code and discussing with colleagues (@sabw8217), we agreed that it feels "right" for the code to be idempotent and handle the idea that an IAM Role might already exist.. there are a number of race conditions in these eventually-consistent APIs that AWS provides. It makes more sense to break the `CreateRole` function out into a few different pieces so that each one is logically testable and clean.

I have changed the interface from the controller from `pkg.awapi.iam.CreateRole` to `pkg.aws.iam.EnsureRole`. That function now makes a call to a new function `GetOrCreateRole`, and that function in turn makes calls to `GetRole` and `CreateRole`. Because we now make the call to `GetRole` _before_ the `CreateCall`, we have a response object that we can use to populate the `RoleARN` and `RoleID`.

**Testing**

I have tested this and been able to replicate both the normal-case (no race condition) and the race-condition case.

**No Race Condition Logs**
```
2021-07-02T00:36:15.359Z        DEBUG   awsapi.iam.EnsureRole   Verifying that IAM Role exists  {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.359Z        DEBUG   awsapi.iam.GetRole      Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.458Z        INFO    awsapi.iam.GetOrCreateRole      Role does not already exist     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.458Z        DEBUG   awsapi.iam.CreateRole   Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.607Z        DEBUG   awsapi.iam.CreateRole   Successfully created the role   {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.607Z        DEBUG   awsapi.iam.EnsureRole   Verifying Tags  {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.607Z        DEBUG   awsapi.iam.VerifyTags   Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.710Z        DEBUG   awsapi.iam.EnsureRole   Attaching Tag   {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.711Z        DEBUG   awsapi.iam.TagRole      Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.836Z        DEBUG   awsapi.iam.TagRole      Successfully completed TagRole call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.836Z        DEBUG   awsapi.iam.EnsureRole   Attaching Permission Boundary   {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.836Z        DEBUG   awsapi.iam.AddPermissionBoundary        Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.939Z        DEBUG   awsapi.iam.AddPermissionBoundary        Successfuly added permission boundary   {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.939Z        DEBUG   awsapi.iam.EnsureRole   Attaching Managed policies      {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:15.939Z        DEBUG   awsapi.iam.AttachManagedRolePolicy      Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app", "policyName": "arn:aws:iam::123456:policy/EKS-us1-stage2-IamManagerController-T5N293MP8WY-default-policy"}
2021-07-02T00:36:16.087Z        DEBUG   awsapi.iam.AttachManagedRolePolicy      Successfully attached managed policies  {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app", "policyName": "arn:aws:iam::123456:policy/EKS-us1-stage2-IamManagerController-T5N293MP8WY-default-policy"}
2021-07-02T00:36:16.087Z        DEBUG   awsapi.iam.EnsureRole   Attaching Inline role policies  {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:16.087Z        DEBUG   awsapi.iam.UpdateRole   Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:16.200Z        DEBUG   awsapi.iam.UpdateRole   Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app", "api": "UpdateAssumeRolePolicy"}
2021-07-02T00:36:16.334Z        DEBUG   awsapi.iam.UpdateRole   AssumeRole Policy is successfully updated       {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:16.335Z        DEBUG   awsapi.iam.AttachInlineRolePolicy       Initiating api call     {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:36:16.478Z        DEBUG   awsapi.iam.AttachInlineRolePolicy       Successfully completed attaching InlineRolePolicy       {"request_id": "2b70bed9-f271-4fdd-998e-8601dbf0dacc", "roleName": "k8s-us1-myns-testrole-app"}
```

**Race Condition Logs**
```
2021-07-02T00:38:38.493Z        DEBUG   awsapi.iam.EnsureRole   Verifying that IAM Role exists  {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.493Z        DEBUG   awsapi.iam.GetRole      Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.608Z        DEBUG   awsapi.iam.GetRole      Successfully able to get the role       {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.608Z        DEBUG   awsapi.iam.EnsureRole   Verifying Tags  {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.608Z        DEBUG   awsapi.iam.VerifyTags   Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.714Z        DEBUG   awsapi.iam.EnsureRole   Attaching Tag   {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.714Z        DEBUG   awsapi.iam.TagRole      Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.865Z        DEBUG   awsapi.iam.TagRole      Successfully completed TagRole call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.865Z        DEBUG   awsapi.iam.EnsureRole   Attaching Permission Boundary   {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.865Z        DEBUG   awsapi.iam.AddPermissionBoundary        Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.970Z        DEBUG   awsapi.iam.AddPermissionBoundary        Successfuly added permission boundary   {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.970Z        DEBUG   awsapi.iam.EnsureRole   Attaching Managed policies      {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:38.970Z        DEBUG   awsapi.iam.AttachManagedRolePolicy      Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app", "policyName": "arn:aws:iam::123456:policy/EKS-us1-stage2-IamManagerController-T5N293MP8WY-default-policy"}
2021-07-02T00:38:39.089Z        DEBUG   awsapi.iam.AttachManagedRolePolicy      Successfully attached managed policies  {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app", "policyName": "arn:aws:iam::123456:policy/EKS-us1-stage2-IamManagerController-T5N293MP8WY-default-policy"}
2021-07-02T00:38:39.089Z        DEBUG   awsapi.iam.EnsureRole   Attaching Inline role policies  {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:39.089Z        DEBUG   awsapi.iam.UpdateRole   Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:39.198Z        DEBUG   awsapi.iam.UpdateRole   Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app", "api": "UpdateAssumeRolePolicy"}
2021-07-02T00:38:39.324Z        DEBUG   awsapi.iam.UpdateRole   AssumeRole Policy is successfully updated       {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:39.324Z        DEBUG   awsapi.iam.AttachInlineRolePolicy       Initiating api call     {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:39.454Z        DEBUG   awsapi.iam.AttachInlineRolePolicy       Successfully completed attaching InlineRolePolicy       {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "roleName": "k8s-us1-myns-testrole-app"}
2021-07-02T00:38:39.454Z        INFO    internal.utils.utils.ParseIRSAAnnotation        Annotation found        {"request_id": "5ee20b83-9c2e-4071-8bd1-51c2b14fd938", "name": "goflask"}
```

### CODE REVIEW

File: .github/DEVELOPER.md [hunk 1]
```diff
@@ -7,6 +7,13 @@
 * Your favorite IDE
 * kubebuilder
 
+### Install Kubebuilder
```
Comment (sabw8217) [inline .github/DEVELOPER.md]: why are you moving this around

File: pkg/awsapi/iam.go [hunk 2]
```diff
@@ -405,27 +419,36 @@ func (i *IAM) AttachInlineRolePolicy(ctx context.Context, req IAMRoleRequest) (*
 	return &IAMRoleResponse{}, nil
 }
 
-//GetRole gets the role from aws iam
-func (i *IAM) GetRole(ctx context.Context, req IAMRoleRequest) (*iam.GetRoleOutput, error) {
-	log := log.Logger(ctx, "awsapi", "iam", "GetRole")
+// CreateRole will try to create an IAM Role, or return back Nil if it does not exist
```
Comment (sabw8217) [inline pkg/awsapi/iam.go]: `or return back Nil if it can not be created`

Comment (diranged) [reply pkg/awsapi/iam.go]: :+1:

File: .github/DEVELOPER.md (hunk 1 above)
Comment (diranged) [reply .github/DEVELOPER.md]: Accidental commit.. i was separately working on tweaking the docs but then I gave up.

**Comment by diranged** [comment]
Closing this - it was merged in https://github.com/keikoproj/iam-manager/pull/82 upstream..

**Pull request closed without merging**